<br/>
<br/>
<br/>
<br/>
 <a href='https://github.com/yanwang271/SpliPath'>SpliPath</a> was developed to nominate novel pathogenic mutation hotspots in genomewide by paried DNA sequencing and RNA splicing data analysis, as described by <a href='file://blabla.html'>Yan et al</a>. It aims to mitigate key challenges of distinguishing aberrant splicing events from artefects and linking rare variants with rare aberrant splicing events. SpliPath provide genomewide splice junction annotation, mutation hotspots nomination, and Shiny-based interactive data visualization, which can be applied to analyse any dataset that includes matched whole genome sequencing and RNAseq data for a common set of donors expected to harbour pathogenic splice-altering rare variants (sRV).
</p>
With the SpliPath data browser, users can visualize extended details on the effects and supporting evidence for detected splice-altering rare variants, and explore heritable aberrant gene splicing events in disease.
<br/>
Please cite this <a href='file://blabla.html'>paper</a> if you use SpliPath in your publication.
<br/>
<br/>
<br/>
<br/>
<br/>
<br/>
