#' Annotate junctions extracted from RNAseq data and filter for junction outliers from LeafCutter analysis result
#'
#' @param rna_meta string RNAseq sample metadata file. The file should contain at least "SampleID", "SubjectID", "Group", "Tissue", and "Path" columns
#' @param chrom string The junctions in this chromosome will be merged among samples.
#' @param tissues_leafcutter_pvals_file list The file paths to the LeafCutter analysis P values of each tissue type. Each element is a LeafCutter file path named under the tissue type. The tissues should be the same with those in the RNAseq sample metadata file.
#' @param max_LeafCutter_Pval numeric The threshold of LeafCutter P value of a novel junction.
#' @param min_nr_tissue numeric Require novel junction LeafCutter P value under the provided threshold in at least min_nr_tissue tissues.
#' @param output_prefix string Prefix of output files.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#' @param addition_annotation list or NULL Optional. Name and file name of additional junction annotation source. Default: list(snaptron = system.file("extdata/Reference/snaptron_gtex_anno.txt", package = "SpliPath")). Whether the input junctions are the given junction list will be shown in the output file (in column "in.annotation_name").
#'
#' @import dplyr
#' @import GenomicRanges
#' @export
annotateLeafCutterJunc <-
  function(rna_meta,
           chrom,
           tissues_leafcutter_pvals_file = NULL,
           max_LeafCutter_Pval = 0.05,
           min_nr_tissue = 1,
           output_prefix, 
           reference = "Default", 
           addition_annotation = list(snaptron = system.file(paste("extdata", "Reference", "snaptron_gtex_anno.txt", sep=.Platform$file.sep), package = "SpliPath"))
  ){
    
    ### Import meta data
    rna_meta = read.table(rna_meta, header = T, sep="\t", stringsAsFactors = F)
    rnaid2subject = lapply(split(rna_meta$SubjectID, rna_meta$SampleID), FUN=unique)
    
    # Check if tissue names exist in meta data
    if (sum(!names(tissues_leafcutter_pvals_file) %in% rna_meta$Tissue) > 0){
      missing_tissue = names(tissues_leafcutter_pvals_file)[!names(tissues_leafcutter_pvals_file) %in% rna_meta$Tissue]
      stop(sprintf("ERROR: Tissue: %s provided in tissues_leafcutter_pvals_file argument not found in meta data", paste(missing_tissue, collapse = ", ")))
    }
    
    ### Merge the junctions
    merged_junc = mergeJuncData(rna_meta, chrom, output_prefix = output_prefix)
    rownames(merged_junc) = paste(merged_junc$chr, merged_junc$start, merged_junc$end, merged_junc$strand, sep=":")
    
    ### Annotate junctions in the junction files of all samples, and then select the novel junctions
    junc_anno = annotateJunc(merged_junc[, c('chr', 'start', 'end', 'strand')], output_prefix, reference = reference, addition_annotation = addition_annotation )

    if (!is.null(addition_annotation)){
      as_annotated = c("in.ensembl", paste0("in.", names(addition_annotation)))
    }else{
      as_annotated = c("in.ensembl")
    }
    junc_anno[, as_annotated] = sapply(junc_anno[, as_annotated] , FUN = as.logical)
    novel_junc = junc_anno[rowSums(junc_anno[, as_annotated]) == 0, ]
    
    ### Merge read count and LeafCutter P values for novel junctions
    if (!is.null(tissues_leafcutter_pvals_file)){
      
      leafcutter_pvals_pass_tissue = NULL
      for (tissue in names(tissues_leafcutter_pvals_file)){
        tissue_sample = rna_meta[rna_meta$Tissue == tissue, "SampleID", drop=T]
        tissue_merged_junc = merged_junc[unique(novel_junc$pos), tissue_sample]
        tissue_merged_junc = reshape2::melt(as.matrix(tissue_merged_junc), varnames = c("Junc", "SampleID"), value.name = paste0("Reads_", tissue))
        tissue_merged_junc[, c("Junc", "SampleID")] = sapply( tissue_merged_junc[, c("Junc", "SampleID")], as.character)
        tissue_merged_junc = tissue_merged_junc[tissue_merged_junc[, paste0("Reads_", tissue)] > 0, ]
        tissue_merged_junc$SubjectID = unname(unlist(rnaid2subject[tissue_merged_junc$SampleID]))
        tissue_merged_junc = tissue_merged_junc[, c("Junc", "SubjectID", paste0("Reads_", tissue))]
        
        leafcutter_pvals_file = tissues_leafcutter_pvals_file[[tissue]]
        leafcutter_pvals = read.table(leafcutter_pvals_file, header = T, sep="\t", row.names = 1, stringsAsFactors = F, check.names=F)
        colnames(leafcutter_pvals) = do.call(rbind, strsplit(colnames(leafcutter_pvals), split = ".Aligned"))[, 1]
        colnames(leafcutter_pvals) = unname(unlist(rnaid2subject[colnames(leafcutter_pvals)]))
        rename_row = do.call(rbind, strsplit(rownames(leafcutter_pvals), "\\:|\\_"))[, c(1:3,6)]
        colnames(rename_row) = c("chr", "start", "end", "strand")
        rename_row[, "end"] = as.integer(rename_row[, "end"]) - 1
        rownames(leafcutter_pvals) = apply(rename_row, 1, paste, collapse=":")
        
        leafcutter_pvals = leafcutter_pvals[row.names(leafcutter_pvals) %in% novel_junc$pos, ]
        leafcutter_pvals = reshape2::melt(as.matrix(leafcutter_pvals), varnames = c("Junc", "SubjectID"), value.name = paste0("Outlier_P_", tissue))
        leafcutter_pvals[, c("Junc", "SubjectID")] = sapply(leafcutter_pvals[, c("Junc", "SubjectID")], as.character)
        leafcutter_pvals = dplyr::left_join(tissue_merged_junc, leafcutter_pvals, by = c("Junc", "SubjectID"), multiple = "all")
        rm(tissue_merged_junc)
        
        if (is.null(leafcutter_pvals_pass_tissue)){
          leafcutter_pvals_pass_tissue = leafcutter_pvals
        }else{
          leafcutter_pvals_pass_tissue = dplyr::full_join(leafcutter_pvals_pass_tissue, leafcutter_pvals, by = c("SubjectID", "Junc"), multiple = "all")
        }
      }
      print(head(leafcutter_pvals_pass_tissue))
      leafcutter_pvals_pass_tissue_idx = rowSums(leafcutter_pvals_pass_tissue[, paste0("Outlier_P_", names(tissues_leafcutter_pvals_file)) ] < max_LeafCutter_Pval, na.rm=T) >= min_nr_tissue  
      leafcutter_pvals_pass_tissue = leafcutter_pvals_pass_tissue[leafcutter_pvals_pass_tissue_idx, ]   
      print(dim(leafcutter_pvals_pass_tissue))
    }
    
    leafcutter_pvals_pass_tissue = dplyr::left_join(leafcutter_pvals_pass_tissue, novel_junc, by = c("Junc" = "pos"), multiple = "all")
    
    output_file = gzfile(sprintf("%s_leafcutter_outliers.txt.gz", output_prefix), 'w')
    write.table(leafcutter_pvals_pass_tissue, output_file, row.names = F, col.names = T, sep = "\t", quote = F)
    close(output_file)
    
  }

