#' Annotate junctions extracted from RNAseq data
#'
#' @param rna_meta string RNAseq sample metadata file. The file should contain at least "SampleID", "SubjectID", "Group", "Tissue", and "Path" columns
#' @param chrom string The junctions in this chromosome will be merged among samples.
#' @param output_prefix string Prefix of output file. The annotations of the junctions will be written in a prefix_anno.txt.gz file.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#' @param addition_annotation list or NULL Optional. Name and file name of additional junction annotation source. Default: list(snaptron = system.file("extdata/Reference/snaptron_gtex_anno.txt", package = "SpliPath")). Whether the input junctions are the given junction list will be shown in the output file (in column "in.annotation_name").
#'
#' @import dplyr
#' @import GenomicRanges
#' @export
annotateJunc <-
function(rna_meta, chrom, output_prefix, reference = "Default", 
         addition_annotation = list(snaptron = system.file(paste("extdata", "Reference", "snaptron_gtex_anno.txt", sep=.Platform$file.sep), package = "SpliPath")) ){
  
  options(scipen = 999)
  if (reference == "Default"){
    gene_ref = system.file(paste("extdata", "Reference", "Gene_GRCh38.98.bed", sep=.Platform$file.sep), package = "SpliPath")
    intron_coding = system.file(paste("extdata", "Reference", "Intron_GRCh38.98.proteincoding.bed", sep=.Platform$file.sep), package = "SpliPath") 
    intron_noncoding = system.file(paste("extdata", "Reference", "Intron_GRCh38.98.noncoding.bed", sep=.Platform$file.sep), package = "SpliPath") 
    merged_junc = "example_chr1.txt.gz"
    
  }else{
    gene_ref = paste(reference, "Gene.bed", sep=.Platform$file.sep)
    intron_coding = paste(reference, "Intron_proteincoding.bed", sep=.Platform$file.sep)
    intron_noncoding = paste(reference, "Intron_noncoding.bed", sep=.Platform$file.sep)
  }
  gene_ref = read.table(gene_ref, sep='\t', header=F, stringsAsFactors = F) 
  colnames(gene_ref) = c('chr',  'start', 'end', 'gene.id', 'gene.name', 'strand', 'gene.type')
  protein_coding_gene = gene_ref[gene_ref$gene.type == "protein_coding", "gene.id"]
  gene_ref = gene_ref[, colnames(gene_ref)!="gene.type"]
  
  intron_coding = read.table(intron_coding, sep='\t', header=F, stringsAsFactors = F) 
  colnames(intron_coding) = c('chr',  'start', 'end', 'transcript.id',  'gene.id', 'strand')
  intron_coding$pos = paste(intron_coding$chr, intron_coding$start, intron_coding$end, intron_coding$strand, intron_coding$gene.id, sep=':')
  intron_coding$start.pos = paste(intron_coding$chr, intron_coding$start, intron_coding$strand, intron_coding$gene.id, sep=':')
  intron_coding$end.pos = paste(intron_coding$chr, intron_coding$end, intron_coding$strand, intron_coding$gene.id, sep=':')
  
  intron_noncoding = read.table(intron_noncoding, sep='\t', header=F, stringsAsFactors = F)
  colnames(intron_noncoding) = c('chr',  'start', 'end', 'transcript.id',  'gene.id', 'strand', 'biotype')
  intron_noncoding$pos = paste(intron_noncoding$chr, intron_noncoding$start, intron_noncoding$end, intron_noncoding$strand, intron_noncoding$gene.id, sep=':')
  
  ensembl_pos = unique(c(paste(intron_coding$chr, intron_coding$start, intron_coding$end, intron_coding$strand, sep=':'),
                  paste(intron_noncoding$chr, intron_noncoding$start, intron_noncoding$end, intron_noncoding$strand, sep=':')))
  
  # Start of new lines
  ### Merge the junctions
  merged_junc = mergeJuncData(rna_meta, chrom, output_prefix = output_prefix)
  merged_junc = merged_junc[, c('chr', 'start', 'end', 'strand')]
  merged_junc = GenomicRanges::makeGRangesFromDataFrame(merged_junc, keep.extra.columns = T, starts.in.df.are.0based=T)
  gene_ref = GenomicRanges::makeGRangesFromDataFrame(gene_ref, keep.extra.columns = T, starts.in.df.are.0based=T)
  overlaps = GenomicRanges::findOverlaps(merged_junc, gene_ref)
  
  merged_junc = data.frame(merged_junc[queryHits(overlaps)])
  merged_junc[, c("seqnames", "strand")] = sapply(merged_junc[, c("seqnames", "strand")], as.character)
  merged_junc = merged_junc[, c("seqnames", "start", "end", "strand")]
  colnames(merged_junc) = c("chr",  "start", "end", "strand")
  
  gene_ref = data.frame(gene_ref[subjectHits(overlaps)])
  gene_ref[, c("seqnames", "strand")] = sapply(gene_ref[, c("seqnames", "strand")], as.character)
  gene_ref = gene_ref[, c("seqnames", "start", "end", "gene.id", 'gene.name', "strand")]
  colnames(gene_ref) = c("chr.map", "start.map", "end.map", "gene.id", 'gene.name', "strand.map")
  
  junc_map = cbind(merged_junc, gene_ref)
  junc_map[, c("start", "start.map")] = junc_map[, c("start", "start.map")] - 1 ### Make coords 0-based
  rm(merged_junc, gene_ref)
  # End of new lines
  
  junc_map$pos = paste(junc_map$chr, junc_map$start, junc_map$end, junc_map$strand, sep=':')
  length(unique(junc_map$pos[duplicated(junc_map$pos)]))/length(unique(junc_map$pos))
  junc_map$start.pos = paste(junc_map$chr, junc_map$start, junc_map$strand, junc_map$gene.id, sep=':')
  junc_map$end.pos = paste(junc_map$chr, junc_map$end, junc_map$strand, junc_map$gene.id, sep=':')
  
  if (!is.null(addition_annotation)){
    for (addition_anno in names(addition_annotation)){
      addition_anno_pos = read.table(addition_annotation[[addition_anno]], header=F, stringsAsFactors = F)$V1
      anno_colname = paste0("in.", addition_anno)
      junc_map[[anno_colname]] = junc_map$pos %in% addition_anno_pos
    }
  }
  junc_map$in.ensembl = junc_map$pos %in% ensembl_pos
  junc_map$coding.gene = junc_map$gene.id %in% protein_coding_gene
  junc_map$coding.transcript = paste(junc_map$pos, junc_map$gene.id, sep= ':') %in% intron_coding$pos
  junc_map$noncoding.transcript = paste(junc_map$pos, junc_map$gene.id, sep=':') %in% intron_noncoding$pos
  
  junc_map$start.anno = junc_map$start.pos %in% intron_coding$start.pos
  junc_map$end.anno = junc_map$end.pos %in% intron_coding$end.pos
  junc_map$start.map = junc_map$start > junc_map$start.map
  junc_map$end.map = junc_map$end < junc_map$end.map
  
  junc_map$map[junc_map$start.map & junc_map$end.map] = "mapped"
  junc_map$map[!junc_map$start.map | !junc_map$end.map] = "d/a_site_unmapped"
  junc_map$map[!junc_map$start.map & !junc_map$end.map] = "unmapped"
  
  junc_map$event = 'annotated'
  junc_map$event[!junc_map$start.anno & junc_map$end.anno & (junc_map$strand == "+")] = "novel_donor"
  junc_map$event[!junc_map$start.anno & junc_map$end.anno & (junc_map$strand == "-")] = "novel_acceptor"
  junc_map$event[junc_map$start.anno & !junc_map$end.anno & (junc_map$strand == "+")] = "novel_acceptor"
  junc_map$event[junc_map$start.anno & !junc_map$end.anno & (junc_map$strand == "-")] = "novel_donor"
  junc_map$event[junc_map$start.anno & junc_map$end.anno & !junc_map$coding.transcript & !junc_map$noncoding.transcript] = "exon_skipping"
  junc_map$event[!junc_map$coding.transcript & junc_map$noncoding.transcript] = "noncoding_transcript"
  junc_map$event[!junc_map$start.anno & !junc_map$end.anno & !junc_map$noncoding.transcript & !junc_map$in.ensembl] = "novel_d&a_site"
  
  exon_skipping_idx = which(junc_map$event == "exon_skipping")
  exon_skipping_anno = lapply(exon_skipping_idx, FUN = function(idx, junc_map){
    ref_intron_gene = intron_coding[intron_coding$gene.id == junc_map[idx, "gene.id"],]
    start = junc_map[idx, "start"]
    end = junc_map[idx, "end"]
    start_transcript = ref_intron_gene[ref_intron_gene$start == start, "transcript.id"]
    end_transcript = ref_intron_gene[ref_intron_gene$end == end, "transcript.id"]
    span_exon_start = TRUE %in% ((ref_intron_gene$end > start) & (ref_intron_gene$end < end))
    span_exon_end = TRUE %in% ((ref_intron_gene$start > start) & (ref_intron_gene$start < end))
    
    if ((length(intersect(start_transcript, end_transcript)) > 0) & (span_exon_start & span_exon_end)){
      return("exon_skipping")
    }else{return("alternative_intron")}
  },
  junc_map = junc_map
  )
  exon_skipping_anno = unlist(exon_skipping_anno)
  junc_map[exon_skipping_idx, "event"] = exon_skipping_anno
  
  # Select junctions mapped within protein-coding genes
  junc_map = junc_map[junc_map$map == "mapped" & junc_map$coding.gene & junc_map$event != "novel_d&a_site",]
  
  # To remove the less likely map when junctions were mapped to multiple genes
  junc_map = junc_map[!(junc_map$coding.transcript == FALSE & junc_map$noncoding.transcript == FALSE & junc_map$in.ensembl == TRUE),]
  
  output_file = gzfile(sprintf("%s_anno.txt.gz", output_prefix), 'w')
  write.table(junc_map, output_file, row.names = F, col.names = T, sep = "\t", quote = F)
  close(output_file)
}

