#' Annotate junctions extracted from RNAseq data
#'
#' @param junc_coordinates data.frame Coordinates of junctions to be annotated. It should contains 4 columns: chromosome, start site, end site, and strand. 
#' @param output_prefix string Prefix of output file. The annotations of the junctions will be written in a prefix_anno.txt.gz file.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#' @param addition_annotation list or NULL Optional. Name and file name of additional junction annotation source. Default: list(snaptron = system.file("extdata/Reference/snaptron_gtex_anno.txt", package = "SpliPath")). Whether the input junctions are the given junction list will be shown in the output file (in column "in.annotation_name").
#'
#' @import dplyr
#' @import GenomicRanges
#' @return data.frame One row in the the data.frame is annotations of a unique junction. 
#' @export
annotateJunc <-
function(junc_coordinates, output_prefix, reference = "Default", 
         addition_annotation = list(snaptron = system.file(paste("extdata", "Reference", "snaptron_gtex_anno.txt", sep=.Platform$file.sep), package = "SpliPath")) ){
  
  options(scipen = 999)
  if (reference == "Default"){
    gene_ref = system.file(paste("extdata", "Reference", "Gene_GRCh38.98.bed", sep=.Platform$file.sep), package = "SpliPath")
    intron_coding = system.file(paste("extdata", "Reference", "Intron_GRCh38.98.proteincoding.bed", sep=.Platform$file.sep), package = "SpliPath") 
    intron_noncoding = system.file(paste("extdata", "Reference", "Intron_GRCh38.98.noncoding.bed", sep=.Platform$file.sep), package = "SpliPath") 
    merged_junc = "example_chr1.txt.gz"
    
  }else{
    gene_ref = paste(reference, "Gene.bed", sep=.Platform$file.sep)
    intron_coding = paste(reference, "Intron_proteincoding.bed", sep=.Platform$file.sep)
    intron_noncoding = paste(reference, "Intron_noncoding.bed", sep=.Platform$file.sep)
  }
  gene_ref = read.table(gene_ref, sep='\t', header=F, stringsAsFactors = F) 
  colnames(gene_ref) = c('chr',  'start', 'end', 'gene.id', 'gene.name', 'strand', 'gene.type')
  protein_coding_gene = gene_ref[gene_ref$gene.type == "protein_coding", "gene.id"]
  gene_ref = gene_ref[, colnames(gene_ref)!="gene.type"]
  
  intron_coding = read.table(intron_coding, sep='\t', header=F, stringsAsFactors = F) 
  colnames(intron_coding) = c('chr',  'start', 'end', 'transcript.id',  'gene.id', 'strand')
  intron_coding$pos = paste(intron_coding$chr, intron_coding$start, intron_coding$end, intron_coding$strand, intron_coding$gene.id, sep=':')
  intron_coding$start.pos = paste(intron_coding$chr, intron_coding$start, intron_coding$strand, intron_coding$gene.id, sep=':')
  intron_coding$end.pos = paste(intron_coding$chr, intron_coding$end, intron_coding$strand, intron_coding$gene.id, sep=':')
  
  intron_noncoding = read.table(intron_noncoding, sep='\t', header=F, stringsAsFactors = F)
  colnames(intron_noncoding) = c('chr',  'start', 'end', 'transcript.id',  'gene.id', 'strand', 'biotype')
  intron_noncoding$pos = paste(intron_noncoding$chr, intron_noncoding$start, intron_noncoding$end, intron_noncoding$strand, intron_noncoding$gene.id, sep=':')
  
  ensembl_pos = unique(c(paste(intron_coding$chr, intron_coding$start, intron_coding$end, intron_coding$strand, sep=':'),
                  paste(intron_noncoding$chr, intron_noncoding$start, intron_noncoding$end, intron_noncoding$strand, sep=':')))
  
  # Start of new lines
  
  # Annotate junctions
  junc_map = annotateJunc_(junc_coordinates, gene_ref, ensembl_pos, protein_coding_gene, intron_coding, intron_noncoding, addition_annotation)
  
  output_file = gzfile(sprintf("%s_anno.txt.gz", output_prefix), 'w')
  write.table(junc_map, output_file, row.names = F, col.names = T, sep = "\t", quote = F)
  close(output_file)
  
  junc_map
}

