#' Quantify the expression of novel junctions based the expression of the background junctions
#'
#' @param junc_file string or data.frame The output file of mergeJuncData function, which contain junction-sample read counts table. The first four columns should be the coordinates of each junction: "chr", "start", "end", "strand".
#' @param junc_anno string or data.frame The output file of annotateJunc function, which contain junction annotations.
#' @param annotation vector Annotation sources (should in colnames in junc_anno) that were combined to determine novel junction. A junction not annotated in any given source is treated as novel junction. Default: c("in.ensembl", "in.snaptron")
#' @param rna_meta string or data.frame The meta data for RNAseq samples. It should contain at least 4 columns: SubjectID, SampleID, Group, Tissue.
#' @param gene_expression_file_prefix string The file prefix of the corrected gene expression data, generated by function correctGeneExp.
#' @param output_prefix string Prefix of output PSI file. The junction-sample PSI table will be written in a prefix_norm.txt.gz file.
#'
#' @return data.frame Seven-column dataframe. Contain junction coordinates, mapped gene, and splicing event of each novel junction.
#' @export
getNovelJuncExp <-
  function(junc_file, junc_anno, annotation = c("in.ensembl", "in.snaptron"), rna_meta, gene_expression_file_prefix, output_prefix){
    if (is.character(junc_file)){
      junc_file = read.table(junc_file, sep = "\t", header = T, stringsAsFactors = F, check.names = F)
    }
    if (is.character(junc_anno)){
      junc_anno = read.table(junc_anno, sep = "\t", header = T, stringsAsFactors = F)
    }
    
    ### Find and calculate the expression of background junctions
    bg_junc = findBackGroundJunc(junc_file, junc_anno, output_prefix, annotation, return_value = T )
    row.names(bg_junc) = paste(bg_junc$chr, bg_junc$start, bg_junc$end, bg_junc$strand, bg_junc$gene.id, sep=":")
    
    # read_count <TO> junc_file
    row.names(junc_file) = paste(junc_file$chr, junc_file$start, junc_file$end, junc_file$strand, sep=":")
    junc_file = junc_file[paste(bg_junc$chr, bg_junc$start, bg_junc$end, bg_junc$strand, sep=":"), ]
    rna_samples = colnames(junc_file)[5:ncol(junc_file)]
    
    ### RNA sample metadata
    rna_meta = read.table(rna_meta, header=T, sep="\t", stringsAsFactors = F)
    
    ### Get binom P for each novel junction 
    calPbinom = function(r, x, trials){
      prob = mean(unlist(tissue_read_count[r, ]) / unlist(tissue_bg[r, ]), na.rm=T) # sum(unlist(x[r, ])) / sum(unlist(trials[r, ]))
      if (!is.na(prob)){
        pbinom(q = unname(unlist(x[r, ])) - 1, 
               size = unname(unlist(trials[r, ])), 
               prob = prob, 
               lower.tail = F)
      }else{
        rep(NA, ncol(x))
      }
    }
    
    ### Get novel junctions expression in samples of one tissue
    tissueRC2Pbinom=function(tissue){ #
      tissue_rna_sample = rna_meta[rna_meta$Tissue == tissue, "SampleID"]
      
      exp_z = read.table(sprintf("%s_%s_gene_expression_zscore.txt", gene_expression_file_prefix, tissue), header=T, row.names=1, stringsAsFactors = F, check.names = F)
      tissue_exp_z = exp_z[bg_junc$gene.id, tissue_rna_sample]
      
      tissue_read_count = junc_file[, tissue_rna_sample]
      
      tissue_bg = bg_junc[, tissue_rna_sample]
      
      tissue_pbinom = data.frame(do.call(rbind, lapply(1:nrow(tissue_bg), FUN=calPbinom, x = tissue_read_count, trials = tissue_bg)), stringsAsFactors = F)
      colnames(tissue_pbinom) = colnames(tissue_bg)
      rownames(tissue_pbinom) = rownames(tissue_bg)
      
      tissue_map_idx = which(tissue_read_count > 0, arr.ind = T)
      
      junc_ls = do.call(rbind, strsplit(row.names(tissue_bg)[unname(tissue_map_idx[,"row"])], split = "[.]"))[,1]
      sample_ls = colnames(tissue_bg)[unname(tissue_map_idx[,"col"])]
      tissue_map = data.frame(Subject=rna_meta[match(sample_ls, rna_meta$SampleID), "SubjectID"],
                              Coordinates_of_novel_junc=junc_ls,
                              Read=tissue_read_count[tissue_map_idx],
                              BG=tissue_bg[tissue_map_idx],
                              P_binom=tissue_pbinom[tissue_map_idx],
                              Gene_expression_Z = tissue_exp_z[tissue_map_idx]
      )
      tissue_map$PSI = round(tissue_map$Read / tissue_map$BG, 4)
      
      output_gz = gzfile(sprintf("%s_novel_junction_exp_%s.txt.gz", output_prefix, tissue), "w")
      write.table(tissue_map[, c("Subject", "Coordinates_of_novel_junc", "Read", "BG", "PSI", "P_binom", "Gene_expression_Z")], output_gz, sep="\t", quote=F, row.names = F)
      close(output_gz)
    }
    
    for (tissue in unique(rna_meta$Tissue)){
      tissueRC2Pbinom(tissue)
    }
  }


