#' Find the variants mapping regions where the possible cis-acting splice-altering variants lie in
#' 
#' The variants mapping region of a novel junction is the disrupted intron and flanking exons.
#' 
#' @param novel_junc dataframe The returned dataframe of callJuncPSI function. Contain coordinates, mapped gene, and splicing event of each novel junction.
#' @param output_prefix string Prefix of output file. The variant mapping region table will be written in a prefix_variant_region.txt.gz file.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#' 
#' @return data.frame Each row is a novel junction and the coordinates of region (column "region.start" and "region.end") where the possible cis-acting splice-altering variants lie in.
#'
#' @export
juncVariantRegion <-
function(novel_junc, output_prefix, reference = "Default"){
  
  if (reference == "Default"){
    ref_exon = system.file(paste("extdata", "Reference", "Exon_GRCh38.98.proteincoding.bed", sep=.Platform$file.sep), package = "SpliPath")
  }else{
    ref_exon = sprintf("%s%sExon_proteincoding.bed", .Platform$file.sep, reference)
  }
  
  exon = read.table(ref_exon, sep="\t", header = FALSE, stringsAsFactors = F) 
  colnames(exon) = c("chr", "start", "end", "exon.id", "gene.id", "strand")
  
  novel_junc$junc = paste(novel_junc$chr, novel_junc$start, novel_junc$end, novel_junc$strand, novel_junc$gene.id, sep=":")
  rownames(novel_junc) = novel_junc$junc
  
  var_region = lapply(novel_junc$junc, FUN = juncVariantRegion_, novel_junc=novel_junc, exon=exon)
  var_region = data.frame(do.call(rbind, var_region))
  colnames(var_region)[9:10] = c("region.start", "region.end")
  
  output_file = gzfile(sprintf("%s_variant_region.txt.gz", output_prefix), "w")
  write.table(var_region, output_file, sep="\t", quote=F, row.names = F)
  close(output_file)
  
  var_region
}
