#' Map potential splice-altering DNA variants to novel junctions in tissues
#' 
#' For each individual, map the DNA variants that of splice-altering potential to a novel junction in this individual's tissues.
#'
#' @param junc_anno_file string or data.frame The output file of annotateJunc function, which contain junction annotations.
#' @param read_count_file string The output file of annotateJunc funciton, which contain junction-sample read count table.
#' @param tissues_leafcutter_pvals_file list The file paths to the LeafCutter analysis P values of each tissue type. Each element is a LeafCutter file path named under the tissue type. The tissues should be the same with those in the RNAseq sample metadata file.
#' @param gdb_path string The path to GDB database that store the genotype data from whole genome sequencing.
#' @param cohort_name string The table name in GDB database that contain sample metadata
#' @param rna_meta string RNAseq sample metadata file. The file should contain at least "SampleID", "SubjectID", "Group, "Tissue" columns.
#' @param wgs_meta string WGS sample metadata file. The file should contain at least "SampleID", "SubjectID", "Group columns.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#' @param as_annotated vector Annotation sources (should be in colnames in junc_anno) that were combined to determine novel junction. A junction not annotated in any given source is treated as novel junction. Default: c("in.ensembl", "in.snaptron")
#' @param max_LeafCutter_Pval numeric Maximum LeafCutter P value of a novel junction to nominate ultra rare sQTL candidate.
#' @param min_nr_tissue numeric Require novel junction LeafCutter P value under the provided threshold in at least min_nr_tissue tissues.
#' @param source_allele_freq list or string Source of the variants allele frequency. Default: "cohort", the allele frequency in the cohort provided by the argument "cohort_name". Or use other allele frequency annotations of the variants recorded in GDB tables, e.g. list(gdb_table = "gnomAD", af_field = "AF"). The gdb_table and af_field are the name of variant annotation table and field where the allele frequency is in the GDB.  
#' @param max_allele_freq numeric Maximum allele frequency to nominate ultra rare sQTL candidate.
#' @param splice_prediction list The splice-altering prediction to annotate the DNA variants. Default: list(SpliceAI = c("spliceaiDS_AG", "spliceaiDS_AL", "spliceaiDS_DG", "spliceaiDS_DL"), dbscSNV = c("rf_score", "ada_score")). The names of elements in the list are prediction tools, they should be in GDB database tables. The items are the scoring fields in the GDB database prediction tables. The maximum of the given scores will be used as the prediction of the corresponding tool for each variants. E.g. The SpliceAI score of a variant will be the maximum of spliceaiDS_AG, "pliceaiDS_AL, spliceaiDS_DG, and spliceaiDS_DL. 
#' @param min_splice_score list The min splice score of each prediction tools to nominate ultra rare sQTL candidate. A min score for SpliceAI must be in the list. Default: list(SpliceAI = 0.2, dbscSNV = 0.7).The predictions of splice sites that score higher than the minimum SpliceAI score will be matched with observed junctions.
#' @param max_skip_exon integer If a novel junction skips more exons than the provided number, it will be considered as "not match" to prediction. Default to 2.  
#' @param SpliceAI_default_reference logical Set to TRUE if the SpliceAI predictions are pre-computed or SpliceAI default annotation were used when running SpliceAI (-A GRCh38, or -A GRCh37). Default to FALSE, indicating the annotation GTF are the same for junction annotation and SpliceAI prediction. 
#' @param blacklist string or NULL Optional. A genomic region blacklist. If provided, the regions will be removed.  Default: system.file("extdata/Reference/blacklist/hg38-blacklist-nochr.v2.bed.gz", package="SpliPath").
#' @param output_prefix string The mapping results will be written in output_prefix_map.txt.gz
#'
#' @import rvat
#' @import RSQLite
#' @import GenomicRanges
#' @import dplyr
#' @import tidyr
#' @returns data.frame One row in the the data.frame is an annotated DNA variant that mapped to a novel junction in a subject's tissues. The junction LeafCutter outlier P value and read count in tissues are in columns "LeafCutter_pVal_tissue" and "Read_tissue".
#' @export 
mapVariantJunc <-
function(junc_anno_file, read_count_file, tissues_leafcutter_pvals_file, gdb_path, cohort_name, rna_meta, wgs_meta,
         reference = "Default",
         as_annotated = c("in.ensembl", "in.snaptron"),
         max_LeafCutter_Pval = 0.05,
         min_nr_tissue = 1,
         source_allele_freq = "cohort", 
         max_allele_freq = 0.05,
         splice_prediction = list(SpliceAI = c("spliceaiDS_AG", "spliceaiDS_AL", "spliceaiDS_DG", "spliceaiDS_DL"), 
                                dbscSNV = c("rf_score", "ada_score")), 
         min_splice_score = list(SpliceAI = 0.2, dbscSNV = 0.7),
         SpliceAI_default_reference = FALSE,
         max_skip_exon = 2,
         blacklist = system.file(paste("extdata", "Blacklist", "hg38-blacklist-nochr.v2.bed.gz", sep=.Platform$file.sep), package = "SpliPath"),
         output_prefix = ""){

  # if (!is.data.frame(junc_var_region)){
  #   junc_var_region = read.table(junc_var_region, header=T, sep="\t", stringsAsFactors = F, check.names = F)
  # }
 
  .func=function(){
  # TESTING
  rna_meta = "SpliPath-test/example_RNAseq_meta.txt"
  wgs_meta = "SpliPath-test/example_DNA_meta.txt"
  gdb_path = "SpliPath-test/example_rvatData.gdb"
  read_count_file = "SpliPath-test/example_chr1.txt.gz"
  junc_anno = "SpliPath-test/example_chr1_anno.txt.gz"
  junc_var_region = read.table("SpliPath-test/example_chr1_variant_region.txt.gz", header=T, sep="\t", stringsAsFactors = F, check.names = F)
  splice_prediction=list(SpliceAI = c("spliceaiDS_AG", "spliceaiDS_AL", "spliceaiDS_DG", "spliceaiDS_DL"), 
                         dbscSNV = c("rf_score", "ada_score"))
  cohort_name = "pheno"
  allele_freq = "gdb"
  as_annotated = c("in.ensembl", "in.snaptron")
  tissues_leafcutter_pvals_file = list(motor_cortex = "SpliPath-test/leafcutter/example_LeafCutter_motor_cortex_outlier_pVals.txt", 
                                       frontal_cortex = "SpliPath-test/leafcutter/example_LeafCutter_frontal_cortex_outlier_pVals.txt", 
                                       cervical = "SpliPath-test/leafcutter/example_LeafCutter_cervical_outlier_pVals.txt", 
                                       lumbar = "SpliPath-test/leafcutter/example_LeafCutter_lumbar_outlier_pVals.txt")
  }
  
  ### Read-in input files first
  rna_meta = read.table(rna_meta, sep="\t", header = T, stringsAsFactors = F)
  wgs_meta = read.table(wgs_meta, sep="\t", header = T, stringsAsFactors = F)
  wgsid2subject = lapply(split(wgs_meta$SubjectID, wgs_meta$SampleID), FUN=unique)

  mygdb = rvat::gdb(gdb_path)
  
  read_count = read.table(read_count_file, sep="\t", header=T, stringsAsFactors = F, check.names = F)
  row.names(read_count) = paste(read_count$chr, read_count$start, read_count$end, read_count$strand, sep=":")
  
  junc_anno = read.table(junc_anno_file, header=T, sep="\t", stringsAsFactors = F, check.names = F)
  junc_anno[, as_annotated] = sapply(junc_anno[, as_annotated] , FUN = as.logical)
  novel_junc = junc_anno[rowSums(junc_anno[, as_annotated]) == 0, ]
  
  junc_var_region = juncVariantRegion(novel_junc, output_prefix = output_prefix, reference = reference)
  junc_var_region = junc_var_region[, c("chr", "region.start", "region.end", "strand", "junc", "event", "gene.name", "gene.id")]
  colnames(junc_var_region)[2:3] = c("start", "end")
  junc_var_region[, c("start", "end")] = sapply(junc_var_region[, c("start", "end")], FUN = as.numeric)
  junc_var_region = junc_var_region[order(junc_var_region$chr, junc_var_region$start), ]
  
  # Get variants and annotation from GDB
  pred_field = paste(paste0(names(splice_prediction), ".*"), collapse = ", ")
  query = sprintf("select VAR_id, var.CHROM, var.POS, var.REF, var.ALT, %s from var ", pred_field)
  for (item in names(splice_prediction)){
    query = paste0(query, sprintf(" left join %s using ('VAR_id')", item))
  }
  query = paste0(query, sprintf(" where var.CHROM in ('%s') ", paste(c(paste0('chr', unique(junc_var_region$chr)), sub("^chr", "", unique(junc_var_region$chr))), collapse = "', '")))
  
  vars = RSQLite::dbGetQuery(mygdb, query)
  vars[is.na(vars)] = 0
  for (item in names(splice_prediction)){
    vars[, splice_prediction[[item]]] = sapply(vars[, splice_prediction[[item]]], as.numeric)
    vars[[item]] = apply(vars[, splice_prediction[[item]]], 1, max)
    vars = vars[, !colnames(vars) %in% splice_prediction[[item]]]
  }
  vars = vars[apply(vars[, names(splice_prediction)], 1, max) > 0, ]
  
  if (!grepl(junc_var_region$chr[1], 'chr', fixed = T)){
    vars$CHROM = sub("^chr", "", vars$CHROM)
  }else{
    if (!grepl(vars$CHROM[1], 'chr', fixed = T)){
      vars$CHROM = paste0("chr", vars$CHROM)
    }
  }
  vars$DNA_variant = paste0(vars$CHROM, ":g.", vars$POS, vars$REF, ">", vars$ALT)
  vars = vars[, c("CHROM", "POS", names(splice_prediction), "VAR_id", "DNA_variant")]
  vars = vars[order(vars$CHROM, as.integer(vars$POS)), ]

  ### Map variant to aberrant splicing region
  # var2junc = bedr::bedr(engine = "bedtools", 
  #                        params = "-d -t all", 
  #                        input = list(a = vars, b = junc_var_region ), 
  #                        method = "closest", check.chr = FALSE , check.merge=F, check.valid = F)
  vars_gr = GenomicRanges::makeGRangesFromDataFrame(vars, start.field = "POS", end.field = "POS", keep.extra.columns = T )
  junc_var_region_gr = GenomicRanges::makeGRangesFromDataFrame(junc_var_region, starts.in.df.are.0based = T, keep.extra.columns = T )
  overlaps = GenomicRanges::findOverlaps(vars_gr, junc_var_region_gr)
  vars_gr = as.data.frame(vars_gr[queryHits(overlaps)], stringsAsFactors = F, row.names = NULL)
  vars_gr = vars_gr[, c("seqnames", "start", names(splice_prediction), "VAR_id", "DNA_variant")]
  colnames(vars_gr) = c("chr", "pos", names(splice_prediction), "VAR_id", "DNA_variant")
  junc_var_region_gr = as.data.frame(junc_var_region_gr[subjectHits(overlaps)], stringsAsFactors = F, row.names = NULL)
  junc_var_region_gr = junc_var_region_gr[, c("seqnames", "start", "end", "strand", "junc", "event", "gene.name", "gene.id")]
  colnames(junc_var_region_gr) = c("region.chr", "region.start", "region.end", "region.strand", "junc", "event", "gene.name", "gene.id")
  var2junc = cbind(vars_gr, junc_var_region_gr)
  # colnames(var2junc) = c("chr", "pos_1", "pos", names(splice_prediction), "VAR_id", "DNA_variant", 
  #                        "region.chr", "region.start", "region.end", "region.strand", "junc", "event", "gene.name", "gene.id", "dist")
  rm(junc_var_region, vars)
  # var2junc = subset(var2junc, dist == 0, select = -pos_1)
  var2junc$var.region = paste(var2junc$region.chr, var2junc$region.start, var2junc$region.end, sep=":")

  geno = rvat::getGT(mygdb, VAR_id=unique(var2junc$VAR_id), cohort=cohort_name)
  geno = rvat::flipToMinor(geno)
  if (allele_freq == "gdb"){
    AF = data.frame(rvat::getAF(geno))
  }else if (class(allele_freq) == "list"){
    AF = rvat::getAnno(mygdb, VAR_id = unique(var2junc$VAR_id), table = allele_freq[["gdb_table"]], fields = allele_freq[["af_field"]])
  }
  colnames(AF) = c("AF")
  
  geno = as.data.frame(SummarizedExperiment::assays(geno)$GT, check.names=F)
  geno[is.na(geno)] = 0
  geno = geno[, colnames(geno) %in% wgs_meta$SampleID]
  colnames(geno) = unname(unlist(wgsid2subject[colnames(geno)]))
  rm(wgs_meta, wgsid2subject)
  
  # psi = read.table(psi_file, header=T, sep="\t", stringsAsFactors = F, check.names = F)
  # psi = psi[!duplicated(psi[, c("chr", "start", "end", "strand", "gene.id")]), ]
  # rownames(psi) = paste(psi$chr, psi$start, psi$end, psi$strand, psi$gene.id, sep=":")
  # 
  
  
  tissue_map = data.frame()
  for (tissue in names(tissues_leafcutter_pvals_file)){
    
    leafcutter_pvals_file = tissues_leafcutter_pvals_file[[tissue]]
    leafcutter_pvals = read.table(leafcutter_pvals_file, header = T, sep="\t", row.names = 1, stringsAsFactors = F, check.names=F)
    colnames(leafcutter_pvals) = do.call(rbind, strsplit(colnames(leafcutter_pvals), split = ".Aligned"))[, 1]
    rename_row = do.call(rbind, strsplit(rownames(leafcutter_pvals), "\\:|\\_"))[, c(1:3,6)]
    colnames(rename_row) = c("chr", "start", "end", "strand")
    rename_row[, "end"] = as.integer(rename_row[, "end"]) - 1
    rownames(leafcutter_pvals) = apply(rename_row, 1, paste, collapse=":")
    
    if (nrow(tissue_map) == 0){
      tissue_map = tissueVar2Junc(rna_meta, tissue, var2junc, "psi", read_count, leafcutter_pvals, geno, names(splice_prediction))
    }else{
      tissue_map = dplyr::full_join(tissue_map, 
                             tissueVar2Junc(rna_meta, tissue, var2junc, "psi", read_count, leafcutter_pvals, geno,  names(splice_prediction)))
    }
  }
  # tissue_map[is.na(tissue_map)] = 0
  if (nrow(tissue_map) == 0){warning("No variant was mapped to novel junctions")}
  
  # Add AF column
  tissue_map$AF = AF[match(as.character(tissue_map$VAR_id), rownames(AF)), "AF"]
  tissue_map = data.frame(lapply(tissue_map, as.character),  stringsAsFactors=F)
  
  # Match observed junction with SpliceAI prediction
  spliceai_pred_match = matchPredNObs(gdb_path = gdb_path, 
                                      cohort_name = cohort_name,
                                      tissue_map, 
                                      min_SpliceAI_score = min_splice_score$SpliceAI, 
                                      max_skip_exon = max_skip_exon,
                                      reference = reference,
                                      SpliceAI_default_reference = SpliceAI_default_reference)
  
  tissue_map$idx = paste(tissue_map$VAR_id, tissue_map$Coordinates_of_novel_junc, tissue_map$Gene_id, sep=":")
  tissue_map$SpliceAI_pred_match_junction = spliceai_pred_match[match(tissue_map$idx, spliceai_pred_match$idx), "match_by_threshold"]
  tissue_map$SpliceAI_pred_cryptic_exon = spliceai_pred_match[match(tissue_map$idx, spliceai_pred_match$idx), "pred_cryptic_exon"]
  
  tissue_map$sQTL_candidate = F
  tissue_map[(!is.na(tissue_map$SpliceAI) & tissue_map$SpliceAI >= min_splice_score$SpliceAI & 
              !is.na(tissue_map$SpliceAI_pred_match_junction) & tissue_map$SpliceAI_pred_match_junction %in% c("full_match", "partial_match")), "sQTL_candidate"] =T
  for (pred_tool in names(min_splice_score)){
    if (pred_tool != "SpliceAI"){
      tissue_map[!is.na(tissue_map[pred_tool])  &  tissue_map[pred_tool] >= min_splice_score[pred_tool] & is.na(tissue_map$SpliceAI_pred_match_junction), "sQTL_candidate" ] = T
    }
  }
  tissue_map[rowSums(tissue_map[paste0("LeafCutter_pVal_", names(tissues_leafcutter_pvals_file))], na.rm=T) < min_nr_tissue, "sQTL_candidate"] = F  
  tissue_map[tissue_map$AF > max_allele_freq, "sQTL_candidate"] = F
  
  output_file = gzfile(sprintf("%s_crossref.txt.gz", output_prefix), "w")
  write.table(tissue_map, output_file, sep="\t", quote=F, row.names = F)
  close(output_file)
  
  tissue_map
}

