#' Convert genome annotation GTF file to reference gene, intron, exon annotation files
#'
#' @param gtf_file string Ensembl genomic annotation GTF file
#' @param intron_bed string BED file contain coordinates (chromosome, start position, end position, strand), transcript id, and gene id information of introns. This file can be generated by GTFtools (https://www.genemine.org/gtftools.php): python gtftools.py -i introns.bed ensembl.gtf  
#' @param output_dir string The output annotation files will be written into the provided directory. The four output files are "Gene.bed": coordinates of genes; "Exon_proteincoding.bed": exons in protein coding transcripts; "Intron_proteincoding.bed": introns in protein coding transcripts; "Intron_noncoding.bed" introns in noncoding transcripts.
#'
#' @import rtracklayer
#' @export
prepareGenoRef <-
function(gtf_file, intron_bed, output_dir){

  gtf2df = function(gtf_file){
   gtf = rtracklayer::import(gtf_file)
   as.data.frame(gtf)
  }
  
  ref.gtf=gtf2df(gtf_file)
  gene_ref = unique(ref.gtf[(ref.gtf$type == 'gene'), c('seqnames', 'start', 'end', 'gene_id', 'gene_name', 'strand', "gene_biotype") ])
  gene_ref$start = gene_ref$start - 1
  gene_ref = gene_ref[order(gene_ref$seqnames, gene_ref$start), ]
  write.table(gene_ref, paste(output_dir, 'Gene.bed', sep = .Platform$file.sep), sep="\t", row.names = FALSE, col.names=FALSE, quote=FALSE)
  
  proteinCodingTranscript = unique(ref.gtf[(ref.gtf$gene_biotype == 'protein_coding') & (ref.gtf$transcript_biotype == 'protein_coding') & !is.na(ref.gtf$transcript_id), 'transcript_id'])
  intron_ref = read.table(intron_bed, sep='\t', header=F, stringsAsFactors = F)
  write.table(intron_ref[intron_ref[, 4] %in% proteinCodingTranscript, ], 
              paste(output_dir, 'Intron_proteincoding.bed', sep = .Platform$file.sep), sep='\t', row.names = FALSE, col.names=FALSE, quote=FALSE)

  noncodingTranscript = ref.gtf[(ref.gtf$gene_biotype == 'protein_coding') &  (ref.gtf$transcript_biotype != 'protein_coding') & !is.na(ref.gtf$transcript_id), 'transcript_id']
  intron_noncoding = intron_ref[intron_ref[, 4] %in% noncodingTranscript,]
  write.table(intron_noncoding, paste(output_dir, 'Intron_noncoding.bed', sep = .Platform$file.sep), sep='\t', row.names = FALSE, col.names=FALSE, quote=FALSE)
  
  exon_ref = unique(ref.gtf[(ref.gtf$gene_biotype == 'protein_coding') & (ref.gtf$transcript_biotype == 'protein_coding') & !is.na(ref.gtf$exon_id), c('seqnames', 'start', 'end', 'exon_id', 'gene_id', 'strand', 'transcript_biotype') ])
  exon_ref$start = exon_ref$start - 1
  exon_ref = exon_ref[order(exon_ref$seqnames, exon_ref$start), ]
  write.table(exon_ref, paste(output_dir, 'Exon_proteincoding.bed', sep = .Platform$file.sep), sep="\t", row.names = FALSE, col.names=FALSE, quote=FALSE)
  
}
