#' Run SpliPath data browser
#'
#' @param rna_meta string Path to RNAseq sample metadata file. There should be at least "SampleID", "SubjecID", "Tissue", "Group" columns in the file.
#' @param dna_meta string Path to DNA sequencing sample metadata file. There should be at least "SampleID", "SubjecID" columns in the file.
#' @param gdb_path string Path to GDB file that store genomic variants data of the DNA sequencing samples. 
#' @param data_dir string Path to junction data prepared for SpliPath data browser by browserData function. It should be the same with the "output_dir" argument of browserData function.
#' @param reference string The directory to genomic annotation files, which can be generated by prepareGenoRef function. Default to pre-made ensembl annotation version GRCh38.98.
#'
#' @import shiny
#' @export
runDataBroswer <-
function(rna_meta, dna_meta, gdb_path, data_dir, reference = "Default"){
    wd = getwd()
    upload_file <<- list()
    
    if (reference == "Default"){
      upload_file$gene_table <<- read.table(system.file(paste("extdata", "Reference", "Gene_GRCh38.98.bed", sep = .Platform$file.sep), package = "SpliPath"), header=F, sep="\t", stringsAsFactors = F)
      upload_file$exon_table <<- read.table(system.file(paste("extdata", "Reference", "Exon_GRCh38.98.proteincoding.bed", sep = .Platform$file.sep), package = "SpliPath"), header=F, sep="\t", stringsAsFactors = F)
    }else{
      upload_file$gene_table <<- read.table(sprintf(paste(reference, "Gene.bed", sep = .Platform$file.sep), header=F, sep="\t", stringsAsFactors = F))
      upload_file$exon_table <<- read.table(sprintf(paste(reference, "Exon.proteincoding.bed", sep = .Platform$file.sep), header=F, sep="\t", stringsAsFactors = F))
    }
    colnames(upload_file$gene_table) <<- c("chr", "start", "end", "gene.id", "gene.name", "strand", "gene.type")
    colnames(upload_file$exon_table) <<- c("chr", "start", "end", "exon.id", "gene.id", "strand")
    upload_file$gene_table <<- upload_file$gene_table[, c("gene.id", "gene.name")]
    upload_file$exon_table$gene.name <<- upload_file$gene_table[match(upload_file$exon_table$gene.id, upload_file$gene_table$gene.id), "gene.name"]
    
    upload_file$rna_meta <<- read.table(rna_meta, header=T, sep="\t")
    upload_file$tissues <<- levels(as.factor(upload_file$rna_meta$Tissue))
    
    upload_file$wgs_meta <<- read.table(dna_meta, header=T, sep="\t")
    
    upload_file$gdb_path <<- paste(wd, gdb_path, sep=.Platform$file.sep)
    
    upload_file$browser_data_dir <<- paste(wd, data_dir, sep=.Platform$file.sep)
    
    upload_file$gene_splice <<- read.table(paste0(upload_file$browser_data_dir, .Platform$file.sep, "number_of_novel_junction_per_subject.txt"), header=T, sep="\t", stringsAsFactors = F, check.names = F)
    
    upload_file$view_gene_list <<- sapply(strsplit(list.files(path = upload_file$browser_data_dir, pattern = "_intron_count.txt.gz$"), split="_"), "[", 2 )
    
    browser_app = paste(system.file(package = "SpliPath"), "SpliPath_Browser", sep= .Platform$file.sep)
    runApp(browser_app)
}
